From fdb6bbba3323be00f375b6265f43d71b56fcdefb Mon Sep 17 00:00:00 2001
From: Eran Ben-Avi <benavi@marvell.com>
Date: Wed, 31 Jul 2013 19:53:00 +0300
Subject: [PATCH 0924/1825] slic: alp: support lantiq external slic

https://github.com/MISL-EBU-System-SW/misl-windriver.git linux-3.4.69-14t2-read
commit d7bf47d8cf7fa0d95309fc7549e5df543f074030

        - Update README, voice tool and SPI HAL
          to support lantiq external slic module

Change-Id: I673c8f5c9d92005c04d7ff560af5327d3d46731a
Signed-off-by: Eran Ben-Avi <benavi@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/2906
Tested-by: Star_Automation <star@marvell.com>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 arch/arm/mach-armada370/mv_hal_if/mvSysTdm.c    |   16 +++++++-
 arch/arm/mach-armadaxp/mv_hal_if/mvSysTdm.c     |   18 +++++++-
 arch/arm/mach-avantalp/mv_hal_if/mvSysTdm.c     |   24 +++++++++--
 arch/arm/mach-feroceon-kw2/mv_hal_if/mvSysTdm.c |   21 ++++++++--
 arch/arm/plat-armada/mv_hal/spi/mvSpi.c         |   16 ++++++++
 arch/arm/plat-armada/mv_hal/spi/mvSpi.h         |    4 +-
 arch/arm/plat-feroceon/mv_hal/spi/mvSpi.c       |   12 +++--
 arch/arm/plat-feroceon/mv_hal/spi/mvSpi.h       |    3 +-
 tools/voice/lantiq/kernel/Makefile              |    4 ++
 tools/voice/lantiq/kernel/README                |   42 ++------------------
 tools/voice/lantiq/kernel/mv_voice_tool.c       |   47 +++++++++++++++++++++++
 11 files changed, 148 insertions(+), 59 deletions(-)

diff --git a/arch/arm/mach-armada370/mv_hal_if/mvSysTdm.c b/arch/arm/mach-armada370/mv_hal_if/mvSysTdm.c
index db94038..c5b66a2 100644
--- a/arch/arm/mach-armada370/mv_hal_if/mvSysTdm.c
+++ b/arch/arm/mach-armada370/mv_hal_if/mvSysTdm.c
@@ -135,6 +135,8 @@ MV_STATUS mvSysTdmInit(MV_TDM_PARAMS* tdmParams)
 
 MV_VOID mvSysTdmSpiRead(MV_U16 lineId, MV_U8* cmdBuff, MV_U8 cmdSize, MV_U8* dataBuff, MV_U8 dataSize)
 {
+	MV_SPI_TYPE spiType = SPI_TYPE_SLIC_ZARLINK_SILABS;
+
 #if defined(MV_TDM_SUPPORT) && !defined(ZARLINK_SLIC_SUPPORT)
 
 	if((cmdSize > 4) || (dataSize > MAX_DATA_LENGTH))
@@ -147,8 +149,12 @@ MV_VOID mvSysTdmSpiRead(MV_U16 lineId, MV_U8* cmdBuff, MV_U8 cmdSize, MV_U8* dat
 
 #else /* MV_COMM_UNIT_SUPPORT || ZARLINK_SLIC_SUPPORT */
 
+#ifdef LANTIQ_SLIC_SUPPORT
+	spiType = SPI_TYPE_SLIC_LANTIQ;
+#endif
+
 	/* Set SPI parameters(lineId = devId) */
-	mvSpiParamsSet(mvBoardTdmSpiIdGet(), mvBoardTdmSpiCsGet(lineId), SPI_TYPE_SLIC);
+	mvSpiParamsSet(mvBoardTdmSpiIdGet(), mvBoardTdmSpiCsGet(lineId), spiType);
 
 	if(MV_OK != mvSpiWriteThenRead (mvBoardTdmSpiIdGet(), cmdBuff, cmdSize, dataBuff, dataSize, 0))
 		printk("SPI read failed !!!\n");
@@ -171,6 +177,8 @@ MV_VOID mvSysTdmSpiRead(MV_U16 lineId, MV_U8* cmdBuff, MV_U8 cmdSize, MV_U8* dat
 *******************************************************************************/
 MV_VOID mvSysTdmSpiWrite(MV_U16 lineId, MV_U8* cmdBuff, MV_U8 cmdSize, MV_U8* dataBuff, MV_U8 dataSize)
 {
+	MV_SPI_TYPE spiType = SPI_TYPE_SLIC_ZARLINK_SILABS;
+
 #if defined(MV_TDM_SUPPORT) && !defined(ZARLINK_SLIC_SUPPORT)
 
 	if((cmdSize > 3) || (dataSize > MAX_DATA_LENGTH))
@@ -183,8 +191,12 @@ MV_VOID mvSysTdmSpiWrite(MV_U16 lineId, MV_U8* cmdBuff, MV_U8 cmdSize, MV_U8* da
 
 #else /* MV_COMM_UNIT_SUPPORT || ZARLINK_SLIC_SUPPORT */
 
+#ifdef LANTIQ_SLIC_SUPPORT
+	spiType = SPI_TYPE_SLIC_LANTIQ;
+#endif
+
 	/* Set SPI parameters(lineId = devId) */
-	mvSpiParamsSet(mvBoardTdmSpiIdGet(), mvBoardTdmSpiCsGet(lineId), SPI_TYPE_SLIC);
+	mvSpiParamsSet(mvBoardTdmSpiIdGet(), mvBoardTdmSpiCsGet(lineId), spiType);
 
 	if(MV_OK != mvSpiWriteThenWrite (mvBoardTdmSpiIdGet(), cmdBuff, cmdSize, dataBuff, dataSize))
 		printk("SPI write failed !!!\n");
diff --git a/arch/arm/mach-armadaxp/mv_hal_if/mvSysTdm.c b/arch/arm/mach-armadaxp/mv_hal_if/mvSysTdm.c
index 83c9d59..a2dbb56 100644
--- a/arch/arm/mach-armadaxp/mv_hal_if/mvSysTdm.c
+++ b/arch/arm/mach-armadaxp/mv_hal_if/mvSysTdm.c
@@ -126,6 +126,8 @@ MV_STATUS mvSysTdmInit(MV_TDM_PARAMS* tdmParams)
 
 MV_VOID mvSysTdmSpiRead(MV_U16 lineId, MV_U8* cmdBuff, MV_U8 cmdSize, MV_U8* dataBuff, MV_U8 dataSize)
 {
+	MV_SPI_TYPE spiType = SPI_TYPE_SLIC_ZARLINK_SILABS;
+
 #if defined(MV_TDM_SUPPORT) && !defined(ZARLINK_SLIC_SUPPORT)
 
 	if((cmdSize > 4) || (dataSize > MAX_DATA_LENGTH))
@@ -138,8 +140,12 @@ MV_VOID mvSysTdmSpiRead(MV_U16 lineId, MV_U8* cmdBuff, MV_U8 cmdSize, MV_U8* dat
 
 #else /* MV_COMM_UNIT_SUPPORT || ZARLINK_SLIC_SUPPORT */
 
+#ifdef LANTIQ_SLIC_SUPPORT
+	spiType = SPI_TYPE_SLIC_LANTIQ;
+#endif
+
 	/* Set SPI parameters(lineId = devId) */
-	mvSpiParamsSet(0, mvBoardTdmSpiCsGet(lineId), SPI_TYPE_SLIC);
+	mvSpiParamsSet(0, mvBoardTdmSpiCsGet(lineId), spiType);
 	
 	if(MV_OK != mvSpiWriteThenRead (0, cmdBuff, cmdSize, dataBuff, dataSize, 0))
 		printk("SPI read failed !!!\n");
@@ -162,6 +168,8 @@ MV_VOID mvSysTdmSpiRead(MV_U16 lineId, MV_U8* cmdBuff, MV_U8 cmdSize, MV_U8* dat
 *******************************************************************************/
 MV_VOID mvSysTdmSpiWrite(MV_U16 lineId, MV_U8* cmdBuff, MV_U8 cmdSize, MV_U8* dataBuff, MV_U8 dataSize)
 {
+	MV_SPI_TYPE spiType = SPI_TYPE_SLIC_ZARLINK_SILABS;
+
 #if defined(MV_TDM_SUPPORT) && !defined(ZARLINK_SLIC_SUPPORT)
 
 	if((cmdSize > 3) || (dataSize > MAX_DATA_LENGTH))
@@ -173,9 +181,13 @@ MV_VOID mvSysTdmSpiWrite(MV_U16 lineId, MV_U8* cmdBuff, MV_U8 cmdSize, MV_U8* da
 	mvTdmSpiWrite(cmdBuff, cmdSize, dataBuff, dataSize, lineId);
 
 #else /* MV_COMM_UNIT_SUPPORT || ZARLINK_SLIC_SUPPORT */
-	
+
+#ifdef LANTIQ_SLIC_SUPPORT
+	spiType = SPI_TYPE_SLIC_LANTIQ;
+#endif
+
 	/* Set SPI parameters(lineId = devId) */
-	mvSpiParamsSet(0, mvBoardTdmSpiCsGet(lineId), SPI_TYPE_SLIC);
+	mvSpiParamsSet(0, mvBoardTdmSpiCsGet(lineId), spiType);
 
 	if(MV_OK != mvSpiWriteThenWrite (0, cmdBuff, cmdSize, dataBuff, dataSize))
 		printk("SPI write failed !!!\n");
diff --git a/arch/arm/mach-avantalp/mv_hal_if/mvSysTdm.c b/arch/arm/mach-avantalp/mv_hal_if/mvSysTdm.c
index 2aaabe1..4af3120 100644
--- a/arch/arm/mach-avantalp/mv_hal_if/mvSysTdm.c
+++ b/arch/arm/mach-avantalp/mv_hal_if/mvSysTdm.c
@@ -135,7 +135,10 @@ MV_STATUS mvSysTdmInit(MV_TDM_PARAMS* tdmParams)
 
 MV_VOID mvSysTdmSpiRead(MV_U16 lineId, MV_U8* cmdBuff, MV_U8 cmdSize, MV_U8* dataBuff, MV_U8 dataSize)
 {
-#if defined(MV_TDM_SUPPORT) && !defined(ZARLINK_SLIC_SUPPORT) && !defined(SILABS_SLIC_SUPPORT) && !defined(LANTIQ_SLIC_SUPPORT)
+	MV_SPI_TYPE spiType = SPI_TYPE_SLIC_ZARLINK_SILABS;
+
+#if defined(MV_TDM_SUPPORT) && !defined(ZARLINK_SLIC_SUPPORT) && \
+!defined(SILABS_SLIC_SUPPORT) && !defined(LANTIQ_SLIC_SUPPORT)
 
 	if((cmdSize > 4) || (dataSize > MAX_DATA_LENGTH))
 	{
@@ -147,8 +150,12 @@ MV_VOID mvSysTdmSpiRead(MV_U16 lineId, MV_U8* cmdBuff, MV_U8 cmdSize, MV_U8* dat
 
 #else /* MV_COMM_UNIT_SUPPORT || ZARLINK_SLIC_SUPPORT || SILABS_SLIC_SUPPORT || LANTIQ_SLIC_SUPPORT */
 
+#ifdef LANTIQ_SLIC_SUPPORT
+	spiType = SPI_TYPE_SLIC_LANTIQ;
+#endif
+
 	/* Set SPI parameters(lineId = devId) */
-	mvSpiParamsSet(mvBoardTdmSpiIdGet(), mvBoardTdmSpiCsGet(lineId), SPI_TYPE_SLIC);
+	mvSpiParamsSet(mvBoardTdmSpiIdGet(), mvBoardTdmSpiCsGet(lineId), spiType);
 	
 	if (MV_OK != mvSpiWriteThenRead(mvBoardTdmSpiIdGet(), cmdBuff, cmdSize, dataBuff, dataSize, 0))
 		printk("SPI read failed !!!\n");
@@ -171,7 +178,10 @@ MV_VOID mvSysTdmSpiRead(MV_U16 lineId, MV_U8* cmdBuff, MV_U8 cmdSize, MV_U8* dat
 *******************************************************************************/
 MV_VOID mvSysTdmSpiWrite(MV_U16 lineId, MV_U8* cmdBuff, MV_U8 cmdSize, MV_U8* dataBuff, MV_U8 dataSize)
 {
-#if defined(MV_TDM_SUPPORT) && !defined(ZARLINK_SLIC_SUPPORT) && !defined(SILABS_SLIC_SUPPORT) && !defined(LANTIQ_SLIC_SUPPORT)
+	MV_SPI_TYPE spiType = SPI_TYPE_SLIC_ZARLINK_SILABS;
+
+#if defined(MV_TDM_SUPPORT) && !defined(ZARLINK_SLIC_SUPPORT) && \
+!defined(SILABS_SLIC_SUPPORT) && !defined(LANTIQ_SLIC_SUPPORT)
 
 	if((cmdSize > 3) || (dataSize > MAX_DATA_LENGTH))
 	{
@@ -182,9 +192,13 @@ MV_VOID mvSysTdmSpiWrite(MV_U16 lineId, MV_U8* cmdBuff, MV_U8 cmdSize, MV_U8* da
 	mvTdmSpiWrite(cmdBuff, cmdSize, dataBuff, dataSize, lineId);
 
 #else /* MV_COMM_UNIT_SUPPORT || ZARLINK_SLIC_SUPPORT || SILABS_SLIC_SUPPORT || LANTIQ_SLIC_SUPPORT */
-	
+
+#ifdef LANTIQ_SLIC_SUPPORT
+	spiType = SPI_TYPE_SLIC_LANTIQ;
+#endif
+
 	/* Set SPI parameters(lineId = devId) */
-	mvSpiParamsSet(mvBoardTdmSpiIdGet(), mvBoardTdmSpiCsGet(lineId), SPI_TYPE_SLIC);
+	mvSpiParamsSet(mvBoardTdmSpiIdGet(), mvBoardTdmSpiCsGet(lineId), spiType);
 
 	if (MV_OK != mvSpiWriteThenWrite(mvBoardTdmSpiIdGet(), cmdBuff, cmdSize, dataBuff, dataSize))
 		printk("SPI write failed !!!\n");
diff --git a/arch/arm/mach-feroceon-kw2/mv_hal_if/mvSysTdm.c b/arch/arm/mach-feroceon-kw2/mv_hal_if/mvSysTdm.c
index c315f14..37c5872 100644
--- a/arch/arm/mach-feroceon-kw2/mv_hal_if/mvSysTdm.c
+++ b/arch/arm/mach-feroceon-kw2/mv_hal_if/mvSysTdm.c
@@ -135,8 +135,10 @@ MV_STATUS mvSysTdmInit(MV_TDM_PARAMS* tdmParams)
 
 MV_VOID mvSysTdmSpiRead(MV_U16 lineId, MV_U8* cmdBuff, MV_U8 cmdSize, MV_U8* dataBuff, MV_U8 dataSize)
 {
-#if defined(MV_TDM_SUPPORT) && !defined(ZARLINK_SLIC_SUPPORT) && !defined(SILABS_SLIC_SUPPORT) && !defined(LANTIQ_SLIC_SUPPORT)
+	MV_SPI_TYPE spiType = SPI_TYPE_SLIC_ZARLINK_SILABS;
 
+#if defined(MV_TDM_SUPPORT) && !defined(ZARLINK_SLIC_SUPPORT) && \
+!defined(SILABS_SLIC_SUPPORT) && !defined(LANTIQ_SLIC_SUPPORT)
 	if((cmdSize > 4) || (dataSize > MAX_DATA_LENGTH))
 	{
 		mvOsPrintf("Error, exceeded max size of command(%d) or data(%d)\n", cmdSize, dataSize);
@@ -147,8 +149,12 @@ MV_VOID mvSysTdmSpiRead(MV_U16 lineId, MV_U8* cmdBuff, MV_U8 cmdSize, MV_U8* dat
 
 #else /* MV_COMM_UNIT_SUPPORT || ZARLINK_SLIC_SUPPORT || SILABS_SLIC_SUPPORT || LANTIQ_SLIC_SUPPORT */
 
+#ifdef LANTIQ_SLIC_SUPPORT
+	spiType = SPI_TYPE_SLIC_LANTIQ;
+#endif
+
 	/* Set SPI parameters(lineId = devId) */
-	mvSpiParamsSet(mvBoardTdmSpiIdGet(), mvBoardTdmSpiCsGet(lineId), SPI_TYPE_SLIC);
+	mvSpiParamsSet(mvBoardTdmSpiIdGet(), mvBoardTdmSpiCsGet(lineId), spiType);
 
 	if(MV_OK != mvSpiWriteThenRead (mvBoardTdmSpiIdGet(), cmdBuff, cmdSize, dataBuff, dataSize, 0))
 		printk("SPI read failed !!!\n");
@@ -171,7 +177,10 @@ MV_VOID mvSysTdmSpiRead(MV_U16 lineId, MV_U8* cmdBuff, MV_U8 cmdSize, MV_U8* dat
 *******************************************************************************/
 MV_VOID mvSysTdmSpiWrite(MV_U16 lineId, MV_U8* cmdBuff, MV_U8 cmdSize, MV_U8* dataBuff, MV_U8 dataSize)
 {
-#if defined(MV_TDM_SUPPORT) && !defined(ZARLINK_SLIC_SUPPORT) && !defined(SILABS_SLIC_SUPPORT) && !defined(LANTIQ_SLIC_SUPPORT)
+	MV_SPI_TYPE spiType = SPI_TYPE_SLIC_ZARLINK_SILABS;
+
+#if defined(MV_TDM_SUPPORT) && !defined(ZARLINK_SLIC_SUPPORT) && \
+!defined(SILABS_SLIC_SUPPORT) && !defined(LANTIQ_SLIC_SUPPORT)
 
 	if((cmdSize > 3) || (dataSize > MAX_DATA_LENGTH))
 	{
@@ -183,8 +192,12 @@ MV_VOID mvSysTdmSpiWrite(MV_U16 lineId, MV_U8* cmdBuff, MV_U8 cmdSize, MV_U8* da
 
 #else /* MV_COMM_UNIT_SUPPORT || ZARLINK_SLIC_SUPPORT || SILABS_SLIC_SUPPORT || LANTIQ_SLIC_SUPPORT */
 
+#ifdef LANTIQ_SLIC_SUPPORT
+	spiType = SPI_TYPE_SLIC_LANTIQ;
+#endif
+
 	/* Set SPI parameters(lineId = devId) */
-	mvSpiParamsSet(mvBoardTdmSpiIdGet(), mvBoardTdmSpiCsGet(lineId), SPI_TYPE_SLIC);
+	mvSpiParamsSet(mvBoardTdmSpiIdGet(), mvBoardTdmSpiCsGet(lineId), spiType);
 
 	if(MV_OK != mvSpiWriteThenWrite (mvBoardTdmSpiIdGet(), cmdBuff, cmdSize, dataBuff, dataSize))
 		printk("SPI write failed !!!\n");
diff --git a/arch/arm/plat-armada/mv_hal/spi/mvSpi.c b/arch/arm/plat-armada/mv_hal/spi/mvSpi.c
index 36c56e6..732876d 100644
--- a/arch/arm/plat-armada/mv_hal/spi/mvSpi.c
+++ b/arch/arm/plat-armada/mv_hal/spi/mvSpi.c
@@ -82,12 +82,20 @@ static MV_SPI_TYPE_INFO spiTypes[] = {
 	{
 		.en16Bit = MV_TRUE,
 		.byteCsAsrt = MV_FALSE,
+		.clockPolLow = MV_TRUE,
 		.baudRate = (20 << 20) /*  20M */
 	},
 	{
 		.en16Bit = MV_FALSE,
+		.clockPolLow = MV_TRUE,
 		.byteCsAsrt = MV_TRUE,
 		.baudRate = _8M
+	},
+	{
+		.en16Bit = MV_FALSE,
+		.clockPolLow = MV_TRUE,
+		.byteCsAsrt = MV_FALSE,
+		.baudRate = _8M
 	}
 };
 
@@ -115,6 +123,8 @@ MV_SPI_TYPE_INFO *currSpiInfo = NULL;
 ********************************************************************************/
 MV_STATUS mvSpiParamsSet(MV_U8 spiId, MV_U8 csId, MV_SPI_TYPE type)
 {
+	MV_SPI_IF_PARAMS ifParams;
+
 	if (csId > (MV_SPI_MAX_CS - 1)) {
 		mvOsPrintf("Error, csId(%d) exceeded maximum(%d)\n", csId, (MV_SPI_MAX_CS - 1));
 		return MV_ERROR;
@@ -130,6 +140,12 @@ MV_STATUS mvSpiParamsSet(MV_U8 spiId, MV_U8 csId, MV_SPI_TYPE type)
 	if (currSpiInfo != (&(spiTypes[type]))) {
 		currSpiInfo = &(spiTypes[type]);
 		mvSpiBaudRateSet(spiId, currSpiInfo->baudRate);
+
+		ifParams.clockPolLow = currSpiInfo->clockPolLow;
+		ifParams.clockPhase = SPI_CLK_BEGIN_CYC;
+		ifParams.txMsbFirst = MV_FALSE;
+		ifParams.rxMsbFirst = MV_FALSE;
+		mvSpiIfConfigSet(spiId, &ifParams);
 	}
 
 	return MV_OK;
diff --git a/arch/arm/plat-armada/mv_hal/spi/mvSpi.h b/arch/arm/plat-armada/mv_hal/spi/mvSpi.h
index d9948e4..d0b031f 100644
--- a/arch/arm/plat-armada/mv_hal/spi/mvSpi.h
+++ b/arch/arm/plat-armada/mv_hal/spi/mvSpi.h
@@ -105,12 +105,14 @@ typedef struct {
 	MV_BOOL	en16Bit;
 	/* should we assert / disassert CS for each byte we read / write */
 	MV_BOOL	byteCsAsrt;
+	MV_BOOL clockPolLow;
 	MV_U32	baudRate;
 } MV_SPI_TYPE_INFO;
 
 typedef enum {
 	SPI_TYPE_FLASH = 0,
-	SPI_TYPE_SLIC
+	SPI_TYPE_SLIC_ZARLINK_SILABS,
+	SPI_TYPE_SLIC_LANTIQ
 } MV_SPI_TYPE;
 
 
diff --git a/arch/arm/plat-feroceon/mv_hal/spi/mvSpi.c b/arch/arm/plat-feroceon/mv_hal/spi/mvSpi.c
index b27084d..e3112ef 100644
--- a/arch/arm/plat-feroceon/mv_hal/spi/mvSpi.c
+++ b/arch/arm/plat-feroceon/mv_hal/spi/mvSpi.c
@@ -67,7 +67,6 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include "mvSpiSpec.h"
 #include "mvSpi.h"
 #include "mvSysSpi.h"
-#include "mvSysTdmConfig.h"
 #include "ctrlEnv/mvCtrlEnvSpec.h"
 
 /* #define MV_DEBUG */
@@ -89,13 +88,16 @@ static MV_SPI_TYPE_INFO spiTypes[] = {
 	{
 		.en16Bit = MV_FALSE,
 		.clockPolLow = MV_TRUE,
-#ifdef LANTIQ_SLIC_SUPPORT
-		.byteCsAsrt = MV_FALSE,
-#else
 		.byteCsAsrt = MV_TRUE,
-#endif
+		.baudRate = _8M,
+	},
+	{
+		.en16Bit = MV_FALSE,
+		.clockPolLow = MV_TRUE,
+		.byteCsAsrt = MV_FALSE,
 		.baudRate = _8M,
 	}
+
 };
 
 MV_SPI_TYPE_INFO *currSpiInfo = NULL;
diff --git a/arch/arm/plat-feroceon/mv_hal/spi/mvSpi.h b/arch/arm/plat-feroceon/mv_hal/spi/mvSpi.h
index 76697d5..d0b031f 100644
--- a/arch/arm/plat-feroceon/mv_hal/spi/mvSpi.h
+++ b/arch/arm/plat-feroceon/mv_hal/spi/mvSpi.h
@@ -111,7 +111,8 @@ typedef struct {
 
 typedef enum {
 	SPI_TYPE_FLASH = 0,
-	SPI_TYPE_SLIC
+	SPI_TYPE_SLIC_ZARLINK_SILABS,
+	SPI_TYPE_SLIC_LANTIQ
 } MV_SPI_TYPE;
 
 
diff --git a/tools/voice/lantiq/kernel/Makefile b/tools/voice/lantiq/kernel/Makefile
index fe432f6..fd917c3 100644
--- a/tools/voice/lantiq/kernel/Makefile
+++ b/tools/voice/lantiq/kernel/Makefile
@@ -14,6 +14,10 @@ ifdef CONFIG_CPU_BIG_ENDIAN
 CFLAGS  += -DCONFIG_CPU_BIG_ENDIAN
 endif
 
+ifeq ($(CONFIG_MV_TDM_USE_DCO),y)
+ CFLAGS += -DMV_TDM_USE_DCO
+endif
+
 CC := $(CROSS_COMPILE)gcc
 LD := $(CROSS_COMPILE)ld
 AR := $(CROSS_COMPILE)ar
diff --git a/tools/voice/lantiq/kernel/README b/tools/voice/lantiq/kernel/README
index ced005d..54a62c8 100644
--- a/tools/voice/lantiq/kernel/README
+++ b/tools/voice/lantiq/kernel/README
@@ -8,56 +8,23 @@
 	CONFIG_LANTIQ_SLIC_SUPPORT=y
 	CONFIG_LANTIQ_SLIC_DUSLIC_XT=y
 	CONFIG_MV_TDM_PCM_CLK_8MHZ=y
-	CONFIG_MV_TDM_USE_INTERNAL_PCLK_SOURCE=y
 
 2. Build Lantiq lib_ifxos-1.5.17:
 
 	$ cd lib_ifxos-1.5.17
-	$ ./configure								\
-		--host=arm-marvell-linux-gnueabi				\
-		--enable-linux-26						\
-		--enable-kernelbuild=<path-to-lsp>				\
-		--enable-kernelincl=<path-to-lsp>/include			\
-		--with-kernel-module
+	$ ./configure --host=arm-marvell-linux-gnueabi --enable-linux-26 --enable-kernelbuild=<path-to-lsp> --enable-kernelincl=<path-to-lsp>/include	--with-kernel-module
 	$ make
 	$ ls -l src/drv_ifxos.ko
 
 3. Build Lantiq drv_tapi-4.10.1.4:
 	$ cd drv_tapi-4.10.1.4
-	$ ./configure								\
-		--host=arm-marvell-linux-gnueabi				\
-		--enable-linux-26						\
-		--with-kernel-incl=<path-to-lsp>/include			\
-		--with-kernel-build=<path-to-lsp>				\
-		--with-kernel-module						\
-		--with-ifxos-incl=<path-to-lib_ifxos-1.5.17>/src/include	\
-		--enable-module							\
-		--enable-tapi4							\
-		--enable-cont-measurement					\
-		--enable-dtmf							\
-		--enable-hsm							\
-		--enable-metering						\
-		--enable-nlt							\
-		--enable-pcm							\
-		--enable-phone-detection					\
-		--enable-voice
+	$ ./configure --host=arm-marvell-linux-gnueabi	--enable-linux-26 --with-kernel-incl=<path-to-lsp>/include --with-kernel-build=<path-to-lsp> --with-kernel-module --with-ifxos-incl=<path-to-lib_ifxos-1.5.17>/src/include --enable-module --enable-tapi4 --enable-cont-measurement --enable-dtmf --enable-hsm	--enable-metering --enable-nlt	--enable-pcm --enable-phone-detection --enable-voice
 	$ make
 	$ ls -l src/drv_tapi.ko
 
 4. Build Lantiq drv_dxt-1.5.3.4:
 	$ cd drv_dxt-1.5.3.4
-	$ ./configure								\
-		--host=arm-marvell-linux-gnueabi				\
-		--enable-linux-26						\
-		--with-kernel-incl=<path-to-lsp>/include			\
-		--with-kernel-build=<path-to-lsp>				\
-		--with-ifxos-incl=<path-to-lib_ifxos-1.5.17>/src/include	\
-		--with-tapi-incl=<path-to-drv_tapi-4.10.1.4>/include		\
-		--enable-cap-measurement					\
-		--enable-cont-measurement					\
-		--enable-lt							\
-		--enable-proc							\
-		--enable-tone-generator
+	$ ./configure --host=arm-marvell-linux-gnueabi	--enable-linux-26 --with-kernel-incl=<path-to-lsp>/include --with-kernel-build=<path-to-lsp> --with-ifxos-incl=<path-to-lib_ifxos-1.5.17>/src/include --with-tapi-incl=<path-to-drv_tapi-4.10.1.4>/include --enable-cap-measurement --enable-cont-measurement	--enable-lt --enable-proc --enable-tone-generator
 	$ make
 	$ ls -l src/drv_dxt.ko
 
@@ -76,8 +43,7 @@
 
 1. Copy all modules and mv_voice_tool to the target.
 
-2. Make device nodes under /dev:
-	# mknod /dev/tdm c 10 74
+2. Make device node under /dev:
 	# mknod /dev/tapi c 253 0
 
 3. Load Lantiq drivers:
diff --git a/tools/voice/lantiq/kernel/mv_voice_tool.c b/tools/voice/lantiq/kernel/mv_voice_tool.c
index a98ab66..bec752c 100644
--- a/tools/voice/lantiq/kernel/mv_voice_tool.c
+++ b/tools/voice/lantiq/kernel/mv_voice_tool.c
@@ -135,6 +135,11 @@ static void slic_digital_loopback(int tdm_fd, unsigned long int iterations);
 static void channel_balancing_test(int tdm_fd, unsigned long int iterations);
 static void wait_for_event(int block);
 static void release(int signum);
+#if defined(MV_TDM_USE_DCO)
+static void set_tdm_clk_config(void);
+static int get_tdm_clk_correction(void);
+static void set_tdm_clk_correction(int correction);
+#endif
 
 int main(void)
 {
@@ -581,6 +586,12 @@ static void sw_tone_test(int tdm_fd, unsigned char line_id)
 			/* Reload timeout */
 			timeout.tv_usec = TIMEOUT;
 		}
+
+		if (ioctl(tdm_fd, TDM_DEV_PCM_STOP, 0)) {
+			printf("Error, unable to stop pcm bus\n");
+			return;
+		}
+
 	}
 }
 
@@ -1066,3 +1077,39 @@ cb_out:
 	for(ch = 0; ch < total_lines; ch++)
 		tapi_line_feed_set(th, TAPI_DEVICE, ch, TAPI_LINE_FEED_STANDBY);
 }
+
+#if defined(MV_TDM_USE_DCO)
+static void set_tdm_clk_config(void)
+{
+	tdm_dev_clk_t tdm_dev_clk;
+
+	/* Config TDM clock */
+	if (ioctl(tdm_fd, TDM_DEV_TDM_CLK_CONFIG, &tdm_dev_clk))
+		printf("%s Error, unable to config TDM clock.\n", TOOL_PREFIX);
+}
+
+static int get_tdm_clk_correction(void)
+{
+	tdm_dev_clk_t tdm_dev_clk;
+
+	/* Get TDM clock */
+	if (ioctl(tdm_fd, TDM_DEV_TDM_CLK_GET, &tdm_dev_clk)) {
+		printf("%s Error, unable to get TDM clock.\n", TOOL_PREFIX);
+		return 0;
+	}
+
+	return tdm_dev_clk.correction;
+}
+
+static void set_tdm_clk_correction(int correction)
+{
+	tdm_dev_clk_t tdm_dev_clk;
+
+	tdm_dev_clk.correction = correction;
+
+	/* Set TDM clock */
+	if (ioctl(tdm_fd, TDM_DEV_TDM_CLK_SET, &tdm_dev_clk))
+		printf("%s Error, unable to set TDM clock.\n", TOOL_PREFIX);
+
+}
+#endif
-- 
1.7.5.4

